### DISCUSSION TITLE
better pattern matching for sequences

### DISCUSSION THREAD

**Discussion by kam1986** [dotnet/fsharp#16217] [Ideas]
I wonder why F# has not generalized the pattern matching over lists to a generalization version for sequences and generated optimized code based on the type use?

it is not uncommon to traverse a collection like you would do with a list to do something depending on the order of the items of the collection appears.

This is not that hard to implement in the compiler since a sequence _s_ like a list either is empty, or has a head and tail which again is a sequence, hence substitute/inline the isEmpty function body with the check for [], and substitute/inline x :: xs with let x = head _s_; let xs = tail _s_. for det specific implementation of isEmpty, head and tail for the used collection. it would lead to a better chance that the compiler can recognize doublications of sub-patterns of the matching and merge them leading to more performant code.

**Comment by kerams** [comment]
https://github.com/fsharp/fslang-suggestions/issues/968#issuecomment-764636595
